### PR TITLE
feat: delete failed resource before creation

### PR DESCRIPTION
**Pull request by mikechesterwang**
context: https://github.com/risingwavelabs/terraform-provider-risingwavecloud/pull/37#discussion_r1544341958

If the creation failed, then the plugin will not update the state, so it can trigger create in the next attempt. 
The next `apply` will try to create a resource again, and before that, it will first delete the failed resource.

### CODE REVIEW

File: internal/provider/resource_privatelink.go [hunk 1]
```diff
@@ -113,9 +113,36 @@ func (r *PrivateLinkResource) Create(ctx context.Context, req resource.CreateReq
...
+		target         = data.Target.ValueString()
+	)
+
+	pl, err := r.client.GetPrivateLinkByName(ctx, connectionName)
+	if err != nil {
+		// ignore not found error
+		if !errors.Is(err, cloudsdk.ErrPrivateLinkNotFound) {
+			resp.Diagnostics.AddError("Get failed", err.Error())
+			return
+		}
```
Comment (wjf3121) [inline internal/provider/resource_privatelink.go:127]: If the Get call failed due to reasons like timeout, does it really make sense to continue to the creation? 🤔

**Review by wjf3121 — COMMENTED** [review]
I have one question about the error on the get call.

Also, should we have test cases reflecting the change?

File: internal/provider/resource_privatelink.go (hunk 1 above)
Comment (mikechesterwang) [reply internal/provider/resource_privatelink.go:127]: a general retry mechanism should be introduced later to fix problems like this.

Comment (wjf3121) [reply internal/provider/resource_privatelink.go:127]: Even if we have try, my point is more of: should we simply stop and throw error or should we continue with creation when we encountered an unknown error? I'm leaning towards the former.

Comment (mikechesterwang) [reply internal/provider/resource_privatelink.go:127]: yes, the current implementation is just throwing an error and stopping the whole procedure.

Comment (wjf3121) [reply internal/provider/resource_privatelink.go:127]: Hmm, did I read something wrong? This is the current implementation with some unrelated code removed:

```go
pl, err := r.client.GetPrivateLinkByName(ctx, connectionName)
if err != nil {
		// Ignore returning errors that signify the resource is no longer existent
		if !errors.Is(err, cloudsdk.ErrClusterNotFound) {
			...
			return
		}
	} else {
		// None error case
	}

// If GetPrivateLinkByName throws a error that is not 404, we reach here
pl, err = r.client.CreatePrivateLinkAwait(...)

```

Looks like we simply proceed with creation the we encountered an unknown error at the Get call?

Comment (mikechesterwang) [reply internal/provider/resource_privatelink.go:127]: as you can see here
```go
if !errors.Is(err, cloudsdk.ErrClusterNotFound) {
  return
}
```
we simply return, not carry on. 

Here is the logic I tend to implement:

if the error is `NotFound`, then proceed.
if the error is not `NotFound`, then return (abort)
if the error is `nil` (meaning there is an existing resource)
 1. delete the resource then proceed if the resource is in failed state
 2. return (abort) as a resource with the same already exist

Comment (wjf3121) [reply internal/provider/resource_privatelink.go:127]: Oh, I missed the negate operation, the logic is perfectly fine in this case. Apology for the confusion🥹

Comment (mikechesterwang) [reply internal/provider/resource_privatelink.go:127]: nah thanks for the thoughtful review, I'm making it clear later

**Review by wjf3121 — APPROVED** [review]
LGTM!

**Pull request merged**